### PR TITLE
fix(deps): support snowflake 3.3.0 (nanoarrow); clean up polars backcompat

### PR DESCRIPTION
**Pull request by cpcloud**
This PR moves our snowflake implementation to work with `snowflake-connector-python` 3.3.0 which uses nanoarrow for data conversion and avoids the long-standing Arrow version constraints, sort of.

I had to add some annoying code to ignore import warnings because the library still warns about incompatible versions of pyarrow despite not actually having this limitation anymore.

### CODE REVIEW

**Comment by cpcloud** [comment]
Clouds are passing:

```
…/ibis on  bump-snowflake is 📦 v7.0.0 via 🐍 v3.10.12 via ❄️  impure (ibis-3.10.12-env)
❯ pytest -m 'bigquery or snowflake' --snapshot-update -n auto --dist loadgroup -q
bringing up nodes...
..........xxxxxxxxxxxxxxxxxxxxx..xxx.sssssssssssss.sssssssss............xx....xx.s....xx............xx.....x..x.........................x..x.....x.xx...x.......x...x.........x.x.xx [  6%]
...................x........x.x...x..x...........x...xx.......xx.xxx.xxxxxxxxxxx..xx.xxx.xx..x..x.xxx............x............................................................x..x.. [ 12%]
.................................................................................................................................x........x.....................x......x.x.......... [ 18%]
x..x.x.....xx..........xx...........x........x.....xx.......x.....xx..............x......x...xxx.x..x...xx...............x.x...xx.xx...xx.....xxx....x.......................x.....x [ 24%]
..x..xx.........x.x........x.xx....x.......x.x...x..............x.........x...x..x.x.........x......x.xx..............x.......xx..x............x....x..xx...xx.....x.x.............. [ 31%]
.x.sx..x.....x..s.....x..x.xxx...x.....x.xx...x...xx.xx.xx....x..xx.......x...s.......x..x..x..x.xx.xx.....x.x...x.x.x..x.x......xxxxx.x.x.xx.x..x..x.xx..xx..xx....x..xx.x.xx..xx.x [ 37%]
.xx.x.x...................xx..x..x.x.x.x.x.x........x...........x....x..........x.x.....x..x.....xxx.xxxx..x....sxx..x.xxx.xx...x..xx.x...xx.x....x...x.....xxx...x.x....x.x.x...... [ 43%]
......x...x.xx..x.......x...xx.x........x.....x..xxxx..x.x.....x...x.......x.......x...x........xx..xxxx........x.x..x....x.......x.x...........xxx.x.xx...........x..x.......x..... [ 49%]
.......x...........x......x......x................x...xx........x...x......x..x................................................x...x................x..x............................ [ 55%]
................x.............x.......xx....x..x.x......x...........x...............x......x..x.xx......x..x............xx..........x.x....x.......x....x...xx....x.x........x.x...x [ 62%]
x...x...............x........xx.....x....x............x...x.......................x............x.....x..................x......................x...............................x...x [ 68%]
.......x.....x.x.x.........xx...xx..x....x.x....xx........xx.xx.x.......xx..x....x.x......x....x..........................................................x.......x......x.......x.. [ 74%]
..........x.............s.......x...............xx.............x.xx.................................................x.....x.............................................x........... [ 80%]
.............................................x..................................................x.........................................x.........................x............... [ 86%]
...x..s.......................x..................................................................................................................................................... [ 93%]
.......................................xxxx...x..x..............................................s................x............................x..x..........................x....... [ 99%]
xx..xxx..x.......                                                                                                                                                                    [100%]
2423 passed, 30 skipped, 444 xfailed in 427.16s (0:07:07)
```

**Comment by gforsyth** [comment]
Not a huge thing, but if we aren't bumping the lower bound to 3.3.0 do we want to suppress all the pyarrow warnings?
Aren't those potentially valid for end users who are using older versions of the connector?

File: ibis/backends/snowflake/__init__.py [hunk 1]
```diff
@@ -35,9 +36,21 @@
...
     AlchemyExprTranslator,
 )
-from ibis.backends.snowflake.converter import SnowflakePandasData
-from ibis.backends.snowflake.datatypes import SnowflakeType
-from ibis.backends.snowflake.registry import operation_registry
+
+with warnings.catch_warnings():
+    if vparse(importlib.metadata.version("snowflake-connector-python")) < vparse(
+        "3.3.0"
+    ):
```
Comment (gforsyth) [inline ibis/backends/snowflake/__init__.py]: Don't we want to only ignore the warnings if we're `>=3.3.0`?

Comment (cpcloud) [reply ibis/backends/snowflake/__init__.py]: Yeah, I just discovered that right as you submitted the comment :)

**Comment by cpcloud** [comment]
Snowflake looks good:

```
…/ibis on  bump-snowflake is 📦 v7.0.0 via 🐍 v3.10.12 via ❄️  impure (ibis-3.10.12-env) took 19s
❯ pytest -m 'snowflake' --snapshot-update -n auto --dist loadgroup -q
bringing up nodes...
...x............x...................x.......................................................................................................x............x...s...................x.. [ 13%]
......x..................x...xx...xx......xx........x...................x...........x........x.............x.........x....x....x.....x..........x.....................x...x......... [ 26%]
xx.x..................x.....x....x...........x.....x.x..........x.....x...xx.....xx......x.x.x.............x...x....x....x...x.........x...x.x..x...x...........................x... [ 40%]
......................x...xxx.xxx.x....x.x..xx..x...x.xx........................x.........x........x..........x.....xxxxx..x....................x.....x.x.......x....x.xx...x....x.. [ 53%]
..x.....x.x.x.x.....x.x.s...x............................x..........x..x...................x....x....x...........x..............................x.........xx...x............x....... [ 66%]
..............x.......x........x.x...x..x.x..........x..xxx............x..xx........xx...xx.xx.....x..x..x...x...x...x...............x....x......x.......xxx..x..x..xx.xx......x..x. [ 80%]
......x...x....x...............x...............xx.....x.........................................................................sxx.x..x..xx.......x..x.xx..xx......x......x..x.xx.. [ 93%]
x.....x.....x..........x.......x..x.xx.x....xx..x.....x..x...............x......xxx..                                                                                                [100%]
1150 passed, 3 skipped, 192 xfailed in 260.35s (0:04:20)
```

**Pull request merged**